### ISSUE TITLE
Rotation on iOS does not look right

### ISSUE THREAD

**Issue by DaveShuckerow** [flutter/flutter#16322]
Internal: b/77275812

From an internal user trying out Flutter on the external flow for iOS:

> I'm trying out flutter for the first time to gauge the quality of the 'native look and feel' to see if it's a platform I should invest my time in learning.
> 
> Using the example hellowworld app from https://flutter.io/get-started/test-drive/#terminal
> 
> See video. I've slowed it down to show more clearly.
> 
> 
> 
> The layout flashes, the text stretches.

**Comment by tvolkert** [comment]
![bff](https://user-images.githubusercontent.com/15253456/38444408-ec06951a-39a3-11e8-8240-913541ff254b.gif)

**Comment by tvolkert** [comment]
the stretching of the aspect ratio is what looks worst here.  We could potentially also cross-fade from a fixed rotated background so that we don't incur the black edges.

**Comment by tvolkert** [comment]
Per @cbracken in #6995 

> Looks like we need to expose better support for `viewWillTransitionToSize:withTransitionCoordinator:` in `FlutterViewController`.

**Comment by adriancmurray** [comment]
@tvolkert 
> > Alternatives include ... allowing some degree of developer control.

I would love some control here. Specifically it would be nice to allow for the device to rotate without utilizing SystemChrome.setPreferredOrientations but not forcing the main widget to repaint/size. I could see instead passing a callback to the WidgetsApp with an orientation change. From a developer's perspective this callback could be passed to something like a ChangeNotifierProvider and allow us to have very granule control of this process. Similar to how the Camera app functions on iOS where the main view doesn't actually switch between states but instead the icons rotate to their appropriate view. This skewing is really really annoying from a UX point of view.

**Comment by dannyvalentesonos** [comment]
@dnfield Any update on this? We are running into this as well, and on iPads it looks pretty bad.
Thanks.

**Comment by dnfield** [comment]
I'm not aware of anyone working on this at the moment - it might take a little bit between holidays coming up and Flutter Interact, but it is being tracked.  @xster may know more.

**Comment by chickenblood** [comment]
The way this rotation is implemented on iOS, is that it generates the final frame by applying an animated size change (not transform) to the entire window content (the root view controller's view), so the layout changes dynamically, while applying a 90-degree rotation transform at the same time. At the end of the rotation, the applied transform and layout match what is should look like in landscape.

I'm not familiar with the Flutter codebase yet, but would be willing to take a look. Would what I describe or something like it be possible with Flutter's layout engine?

**Comment by dnfield** [comment]
We want to implement [-viewWillTransitionToSize:withTransitionCordinator:](https://developer.apple.com/documentation/uikit/uicontentcontainer/1621466-viewwilltransitiontosize?language=objc) in [FlutterViewController](https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm)

With that, we should be able to pipe the relevant information down to the framework so it can know to run an animation for rotation and when that animation should end.  One thing unclear to me is how long that animation should be, and whether we'll be able to communicate with the framework quickly enough to actually get the animation going in time.

Another option would be to just try to do it all in FlutterViewController.mm and basically just do a size animation on whatever our last frame was unti lwe finish rotating.

**Comment by adriancmurray** [comment]
@dnfield in the mean time, is there a known method to hide the transition during the 200 (or so) milliseconds it takes to rotate? I made an attempt by extending WidgetsFlutterBinding, made my own runApp method to pass my own WidgetsBinding, overrode `handleMetricsChanged` to create a ChangeNotifier that would cover the app with a color and fade out quickly after 200ms (just a rough guess). While it was much prettier, it still wasn't fast enough to cover the first frames of the wretched skew. Am I correct in assuming that without making modifications to the engine that I won't be able to overcome the skew?

**Comment by dnfield** [comment]
I believe it will require engine modifications to be able to start drawing on time - not 100% sure though.  @chinmaygarde may know about that?

**Comment by chinmaygarde** [comment]
Making the transition less jarring does not require any help from the engine. Besides, the engine can do little to help because, in the transitionary states, its primary render target is not correctly sized. An implementation that makes these transitions as seamless as possible (within described constraints) is here https://github.com/flutter/engine/pull/14101

**Comment by dnfield** [comment]
By engine modifications I meant modifications to the engine repository as opposed to framework :)

This seems like what I was expecting would be possible.  It'd be really slick if we could let the framework do some custom things on rotation with widgets, but I think your explanation in the linked PR is pretty convincing about why that wouldn't work very well.

**Comment by chinmaygarde** [comment]
The linked PR did not land because of the issues mentioned in the [final comment](https://github.com/flutter/engine/pull/14101#issuecomment-566734062). While the black bars and flashing can be fixed, I haven't been able to figure out how best to deal with the change in aspect ratio without it looking unnatural.

**Comment by adriancmurray** [comment]
Thanks for working on that. While a natural rotation may be the goal, any chance on implementing a small fix like fading between the two orientations? Meaning that you shouldn’t need to worry too much about following the transition and aspect ratio. This is how it’s handled on the iOS 13’s iMessage application. I think fading is far more acceptable than the current skew.

**Comment by mdccxv** [comment]
We're running into this issue as well, despite Landscape orientation being implemented, we're potentially locking the app to Portrait-only as those iOS transitions look super bizarre.
Any update on possible timeline or at least whether or not it will be fixed at all?

**Comment by dannyvalentesonos** [comment]
@dnfield @chinmaygarde Any update on this? I've also just filed #62537 as it causes platform views to "look" like they are leaving their containers when rotating, since those rotate correctly.

Thanks.

**Comment by rufw91** [comment]
What if the degree of rotation in the animation was used to calculate the aspect ratio linearly as opposed to instantly changing the aspect ratio at the start of the rotation animation.

**Comment by joshkahane** [comment]
Not fammiliar enough to know if there would be limitations preventing this; But would it feasible to update the OrientationBuilder widget to provide a pre and post transform rebuild?

This would allows devs to adjust their UI pre transform to help improve the visual of the transition where a singular transform/animation provided by Flutter can only ever be a one size fits all and ultimately varying in effectiveness depending on the UI at hand.

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.22.0-10.0.pre.88` but only on emulator.
<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.22.0-10.0.pre.88, on Mac OS X 10.15.6 19G2021, locale en-GB)
    • Flutter version 1.22.0-10.0.pre.88 at /Users/markus/development/flutter_master
    • Framework revision e10bdbbd9d (18 hours ago), 2020-09-07 11:40:03 -0700
    • Engine revision ac8b9c4c52
    • Dart version 2.10.0 (build 2.10.0-86.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/markus/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (5 available)
    • Nevercode’s iPhone (mobile)         • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.6
    • iPhone SE (2nd generation) (mobile) • E88D23AA-71AC-49A6-A418-058CDE0D2FC4     • ios            • com.apple.CoreSimulator.SimRuntime.iOS-13-6 (simulator)
    • macOS (desktop)                     • macos                                    • darwin-x64     • Mac OS X 10.15.6 19G2021
    • Web Server (web)                    • web-server                               • web-javascript • Flutter Tools
    • Chrome (web)                        • chrome                                   • web-javascript • Google Chrome 85.0.4183.83

• No issues found!
```
</details>

**Comment by SoolyChristy** [comment]
any progress?

**Comment by pcsosinski** [comment]
@kf6gpe can you help with getting this into a milestone?  it's moved up to P3 based on b/77275812

**Comment by kf6gpe** [comment]
Discussed with the engine team; it looks like we'll need help from @xster and his team on this as well. I've reached out to him to see about arranging for a milestone and assignee.

**Comment by jiahaog** [comment]
Internal bug b/77275812 is marked as P2 so marking this issue as P2 as well.

**Comment by kf6gpe** [comment]
P2 for us in Github maps to "blocking customers soon", not upstream P2. This should probably be a P4 --- "on our list but not currently working on it." Downgrading.

**Comment by mehmetf** [comment]
Internal bug b/77275812 is marked as P2 so marking this issue as P2 as well.

**Comment by kf6gpe** [comment]
Downgrading as we're not actively working on this --- it's a P2 for the customer and a bona fide bug, which corresponds to our P4.

**Comment by mehmetf** [comment]
Internal bug b/77275812 is marked as P2 so marking this issue as P2 as well.

**Comment by kf6gpe** [comment]
Per https://github.com/flutter/flutter/wiki/Triage#triaging-issues, we use P2 on Github for issues we're actively working. A P2 issue upstream is a P4, something we'd like to address but aren't currently working on.

**Comment by chingjun** [comment]
Internal bug b/77275812 is marked as P2 so marking this issue as P2 as well.

**Comment by goderbauer** [comment]
@chingjun please see @kf6gpe's comment just above your own comment about why this in fact is not a P2: https://github.com/flutter/flutter/issues/16322#issuecomment-779603936

**Comment by tvolkert** [comment]
Bumping this to P3 so it stands a better chance at getting picked up in our fidelity planning.

**Comment by antonkrasov** [comment]
Hey guys, is there any chance to get some updates here? This is a quite annoying issue for an app with video. Thanks!

**Comment by IlyaMax** [comment]
Is this something which can be fixed at all with flutter? I see there was an attempt but unsuccessful. This is critical for video player apps as mentioned above. And does it have chance to be fixed in the near future?

**Comment by ldhlfzysys** [comment]
I think it's not enough to implement `-viewWillTransitionToSize:withTransitionCordinator:` in `FlutterViewController`

In add2app mode on iOS, FlutterView is not only ViewController usage but also View usage. such as In my application. I add FlutterView to a native view as child view. when I change FlutterView's Size. It will also has this problem.  but not trigger `-viewWillTransitionToSize:withTransitionCordinator:` API.

So. we need a general solution to  FlutterView Size Change. not only roation screen.

**Comment by SoolyChristy** [comment]
how is it going

**Comment by SamerCat** [comment]
still not working

**Comment by snoopdoggy322** [comment]
any updates?

**Comment by danagbemava-nc** [comment]
Reproducible on the latest versions of flutter

<details>
<summary>recording</summary>

https://user-images.githubusercontent.com/88313112/210337656-9c0d4ad8-f7d9-4ec2-9091-e5a0aed21b50.mov
</details>

<details>
<summary>updated sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  SystemChrome.setPreferredOrientations([
    DeviceOrientation.portraitUp,
    DeviceOrientation.landscapeLeft,
    DeviceOrientation.landscapeRight,
  ]);
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Center(
        child: Text('hello world'),
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.3.10, on macOS 13.1 22C65 darwin-arm, locale en-GB)
    • Flutter version 3.3.10 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 135454af32 (3 weeks ago), 2022-12-15 07:36:55 -0800
    • Engine revision 3316dd8728
    • Dart version 2.18.6
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.74.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • 4F72110C-F38B-4CF9-93C4-4D6042148D28 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-1 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.1 22C65 darwin-arm
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 108.0.5359.124
    ! Device adb-5dd3be00-22GXB2._adb-tls-connect._tcp. is offline.

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[!] Flutter (Channel master, 3.7.0-13.0.pre.138, on macOS 13.1 22C65 darwin-arm64, locale en-GB)
    • Flutter version 3.7.0-13.0.pre.138 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b938dc13df (23 hours ago), 2023-01-02 05:47:23 -0500
    • Engine revision 472e34cbbc
    • Dart version 3.0.0 (build 3.0.0-76.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.74.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • 4F72110C-F38B-4CF9-93C4-4D6042148D28 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-1 (simulator)
    • macOS (desktop)        • macos                                • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 108.0.5359.124
    ! Device adb-5dd3be00-22GXB2._adb-tls-connect._tcp. is offline.

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

```
</details>

**Comment by stephen-ullom** [comment]
Seriously? Five years later? I guess I should have just gone native iOS after all.

**Comment by nikuz** [comment]
This issue diminishes any effort you put into your application. Since after all, it looks cheap and buggy on the very first screen rotation. Is there a way to fix this?

**Comment by jmagman** [comment]
Also internally tracked with b/266184881.

**Comment by hellohuanlin** [comment]
This is a summary of our research so far, in case it can be helpful for a potential better solution in the future: 

**1. Snapshot trick**

This was [initially attempted by Chinmay](https://github.com/flutter/engine/pull/14101). On device, we add a snapshot during the rotation, and then fade it out after the rotation is complete. This approach simply moves the distortion from the beginning of the rotation to the end of the rotation, so the same amount of distortion is still there, just moved to the end ([video here](https://www.youtube.com/watch?v=pNvAbmcOdoI&t=21s)). 

**2. Improved snapshot trick** 
We improved this approach by swapping snapshot in the middle of the rotation (rather than at the end of rotation). The benefits are: 
a. it actually reduces the max distortion roughly from ~4x to ~2x
b. the swapping (or "jumping effect") happens in the middle, when it's rotating the fastest, making it harder to notice

However, there are still 2 drawbacks:  
a. it doesn't work well with dynamic content (animation or video)
b. it takes about 0.5 second to take the snapshot

**3. Size interpolation**
The original idea was from [Dan's comment](https://github.com/flutter/flutter/issues/16322#issuecomment-560930492). Basically the engine sends interpolated sizes during the rotation. (PR [here](https://github.com/flutter/engine/pull/40412))

This approach worked pretty well for simple UI - there are almost no distortions at all during the rotation. However, on complex UI (e.g. flutter gallery), we experienced significant frame drops. Reducing the number of interpolations helps with frame drops, but it also resulted in a very un-smooth transition. 

**4. Delayed swap**

This is actually the simplest approach - just delay the swap of width and height until the middle of rotation, rather than at the beginning of the rotation. (PR [here](https://github.com/flutter/engine/pull/40730))

This approach gives us the same benefit as approach 2, without the 2 drawbacks discussed above. Note that there is still a swapping (or "jumping effect") in the middle of rotation, but it's significantly less noticeable than the existing behavior, which "swaps/jumps" at the beginning of rotation, with larger amount of distortion. 

So in conclusion, with the pros and cons discussed above, we will go with **approach 4**.

**Comment by b3nni97** [comment]
@hellohuanlin Couldn't we take a screenshot at the start of the rotation where height and width have already changed and then fade it in?
That's what native iOS does anyway, so the jump in the middle is no longer really noticeable.

**Comment by hellohuanlin** [comment]
@b3nni97 

> Couldn't we take a screenshot at the start of the rotation where height and width have already changed and then fade it in?

When you take a screenshot where height and width already changed, the screenshot itself has the max distortion. So the result won't look good. 

> That's what native iOS does anyway, so the jump in the middle is no longer really noticeable.

Nope. The native iOS actually does size interpolation.

**Comment by b3nni97** [comment]
@hellohuanlin 
Ok maybe I don't understand something correctly but,

in a native iOS app, it doesn't look to me like the size is interpolated.

![IMG_1427 (1)](https://user-images.githubusercontent.com/18357332/230182058-dbefbc7e-0eb8-453e-bb85-f930c210832a.png)


This screenshot was taken at the beginning of a screen rotation. You can see 2 "screenshots" one of the state before the rotation, and one of the state after the rotation, with the exception that the end state didn't just change width / height, but the height remains (so that the complete height can be filled during the fade in).
During the animation, the width of the clipped part becomes smaller and smaller, but the lower part is clipped.

![IMG_1428 (1)](https://user-images.githubusercontent.com/18357332/230182676-5165eeaa-35be-4623-933c-f847965f796a.png)
This screenshot was taken shortly before the rotation is finished. Here you can see that the end state is now almost completely visible.

I hope you understand what I mean.

Here is the full video of the rotation

https://user-images.githubusercontent.com/18357332/230182421-6839770b-0303-4c87-89f3-73a747302f42.MP4

**Comment by hellohuanlin** [comment]
@b3nni97 I don't think apple is taking a screenshot (otherwise there would be distortion of the screenshot when the window's aspect ratio changes). 

It's impossible to take a screenshot of what the screen will look like *after the rotation is complete* at the time before the rotation (if that's what you want) - the screenshot shows exactly what the current screen is when the screenshot is taken. 

Your particular example is interesting - it changes the entire layout to a split view, in which case I'm not sure what apple is doing under the hood. Without layout change, apple's behavior is size interpolation, as shown below: 

https://user-images.githubusercontent.com/41930132/230184036-7384dadd-687f-487b-adc4-03b20063528d.mp4

**Comment by b3nni97** [comment]
@hellohuanlin 
Ok then this problem is more complex than thought.

> It's impossible to take a screenshot of what the screen will look like _after the rotation is complete_, at the time before the rotation even starts (if that's what you want) - the screenshot shows exactly what the current screen is when the screenshot is taken.

Hmm, I know too little about the engine, I thought it would be possible to do a complete rebuild with the new width and height, and then display this as a "screenshot".

**Comment by hellohuanlin** [comment]
> Hmm, I know too little about the engine, I thought it would be possible to do a complete rebuild with the new width and height, and then display this as a "screenshot".

The screenshot is not flutter-related tho. It's just UIKit API to take a screenshot of the current screen. So once the end state is displayed, you can't "blend it in", as it's already on the screen.

**Comment by tvolkert** [comment]
I think @b3nni97 may be suggesting we could build a new tree with the new width & height and paint it to a buffer rather than the screen.

**Comment by hellohuanlin** [comment]
@tvolkert Gotcha. That likely requires big changes in both framework and engine. However, a new idea came to me that **fading in end-state snapshot** should have similar result to **fading out begin-state snapshot**. The latter is much easier to do, so I did a quick experiment (and I call this "**approach 5**"): 

**Approach 5**

Fade out begin-state snapshot during the rotation. Here's the result: 

https://user-images.githubusercontent.com/41930132/230248611-55ae1e66-dcd2-4251-b1a9-c56902b05160.mp4

This doesn't look at good as the original **approach 2** when I tried it on device, because: 

1. it still has max distortion in the beginning, just not fully opaque
2. the overlapping of the 2 states looks bad

There's another approach came to my mind that combines approach 2 and 5, and I call this **approach 6**. 

**Approach 6**: 

Show the begin-state snapshot for 40% of rotation, then fade out for 20% of rotation, then show end-state real screen for 40% of rotation (number adjustable, I just find 40%-20%-40% result in the smoothest transition). Here's the result: 

https://user-images.githubusercontent.com/41930132/230249537-a929d9fa-3f77-466e-9f08-e4f300b42a60.mp4

**Approach 6** is smoother than **approach 2**. I think it has the best result among approaches 1, 2, 5 and 6. 

However, I still think approach 4 is better than any snapshot tricks, since it works for dynamic contents (videos & animations).
